### DISCUSSION TITLE
Problem running HOOMD with AZPLUGINS

### DISCUSSION THREAD

**Discussion by tpaul-softmat** [glotzerlab/hoomd-blue#1930] [Support] [answered]
Hello!

I am new to Hoomd-Blue. I am using HOOMD 4.8.2 on CPU. Recently, I tried to install 'azplugins' (https://github.com/mphowardlab/azplugins) which is supposedly compatible with this version. After a successful installation, I tried to test the installation by the following command, as suggested on the doc page:
_python -m pytest --pyargs hoomd.azplugins_

However, it returns one error. Please look at the attached screenshot of the error message.

![azplugins-test-error](https://github.com/user-attachments/assets/42bf6b73-abe3-4823-a983-ef10bfb2bc23)

I have tried changing the HOOMD version but with no help. Could you please suggest a possible solution to this?

Many thanks for considering my request.
Tanay.

**Comment by tcmoore3** [comment]
Hi Tanay,

It looks like your azplugins install cannot find  your HOOMD libraries.
Without more information, it is impossible to know why. I recommend
reaching out to the azplugins developers (with the CMake output from
building and installing azplugins) for help.

Tim

**Comment by mphoward** [comment]
Are you able to run any of the regular HOOMD unit tests, or do you get the same error?

https://hoomd-blue.readthedocs.io/en/v4.9.1/testing.html#running-tests

**Comment by tpaul-softmat** [comment]
Thank you all for your responses!

As @mphoward asked, I tried running some regular HOOMD unit tests. I was unsure about the build directory, though. So, I am giving all the details and uploading the corresponding snapshots below. For your information, I installed HOOMD using _conda_.

1. I used Python to find where HOOMD is installed by doing the following: 
    > `import hoomd`
    > `print(hoomd.__file__)`

    The output was `anaconda3/lib/python3.11/site-packages/hoomd`. Then I ran `python3 -m pytest` in that directory. It returned the same errors I got while testing for `azplugins`. Below I attach the snapshot of the error messages.

    ![azplugins-hoomd-test-error](https://github.com/user-attachments/assets/d0b3caf2-35dd-4ad2-b3f8-2e2bdd52103d)

2. Next, I checked if HOOMD initializes properly using the following script:
    > `import hoomd`
    > `device = hoomd.device.auto_select()`
    > `sim = hoomd.Simulation(device=device)`

    It ran without error.

3. Next, I navigated to the subdirectory `pytest` in `anaconda3/lib/python3.11/site-packages/hoomd` and ran `pytest`. Among 2868 items, 2846 passed, and 22 were skipped. Please have a look at the snapshot attached below. 

    ![hoomd-pytest](https://github.com/user-attachments/assets/0d90de48-f639-44cf-a51f-afd7eb86250c)

Please let me know whether I have successfully tested the HOOMD installation. If not, then could you let me know how to do that? If HOOMD is correctly installed, please let me know where the problem is and how to rectify it.

Thanks again for your time and assistance. Best regards,
Tanay.

**Reply by joaander** [reply]
For future reference, `python3 -m pytest --pyargs hoomd` [(as mentioned in the documentation)](https://hoomd-blue.readthedocs.io/en/v4.9.1/testing.html#running-tests) will test the HOOMD-blue installation in your current python environment. There is no need to switch directories.

**Reply by tpaul-softmat** [reply]
Many thanks, for the clarification. I tried with `--pyargs` flag. The same errors I got.

**Comment by joaander** [comment]
You posted images of text, not actual text. This makes it frustrating to try to reply as I cannot copy and paste relevant portions of your output. My reply is therefore somewhat vague.

That being said: I have seen this type of error before when an extension module is imported *before* the main `_hoomd.so` which defines the base classes. @mphoward, or one of the other azplugins developers, should check their code and tests to ensure that `import hoomd` appears before any import of an azplugins .so file.

**Reply by tpaul-softmat** [reply]
I apologize for not copying the actual texts. Here, I am copying the errors I got while running tests for `HOOMD` as well as `azplugins`:
```
============================= test session starts ==============================
platform linux -- Python 3.11.8, pytest-7.4.4, pluggy-1.0.0
rootdir: /home/tanay
plugins: anyio-4.6.2
collected 0 items / 1 error                                                    

==================================== ERRORS ====================================
________________________ ERROR collecting test session _________________________
anaconda3/lib/python3.11/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>:1204: in _gcd_import
    ???
<frozen importlib._bootstrap>:1176: in _find_and_load
    ???
<frozen importlib._bootstrap>:1126: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:241: in _call_with_frames_removed
    ???
<frozen importlib._bootstrap>:1204: in _gcd_import
    ???
<frozen importlib._bootstrap>:1176: in _find_and_load
    ???
<frozen importlib._bootstrap>:1147: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:690: in _load_unlocked
    ???
<frozen importlib._bootstrap_external>:940: in exec_module
    ???
<frozen importlib._bootstrap>:241: in _call_with_frames_removed
    ???
anaconda3/lib/python3.11/site-packages/hoomd/azplugins/__init__.py:7: in <module>
    from hoomd.azplugins import bond, flow, pair
anaconda3/lib/python3.11/site-packages/hoomd/azplugins/bond.py:7: in <module>
    from hoomd.azplugins import _azplugins
E   ImportError: generic_type: type "PotentialBondDoubleWell" referenced unknown base type "hoomd::ForceCompute"
=========================== short test summary info ============================
ERROR  - ImportError: generic_type: type "PotentialBondDoubleWell" referenced unknow...
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!

```

Hope this helps.

**Reply by joaander** [reply]
The error message I was referring to is:
`ImportError: generic_type: type "PotentialBondDoubleWell" referenced unknown base type "hoomd::ForceCompute"`.

pybind11 requires that the definition of the base class be imported before any derived classes can be imported.

**Reply by mphoward** [reply]
> I have seen this type of error before when an extension module is imported _before_ the main `_hoomd.so` which defines the base classes. @mphoward, or one of the other azplugins developers, should check their code and tests to ensure that `import hoomd` appears before any import of an azplugins .so file.

There is not an explicit call to `import hoomd` before `_azplugins` is imported:

https://github.com/mphowardlab/azplugins/blob/6de4cb21f4b7a552259744cbb521191ba67543ff/src/bond.py#L7

However, that is also not the case in the example plugins, ex:

https://github.com/glotzerlab/hoomd-blue/blob/trunk-patch/example_plugins/pair_plugin/pair.py#L7

We also follow (mostly) the standard component template, so we are running CI on all commits and none of them has failed with this error, ex:

https://github.com/mphowardlab/azplugins/actions/runs/11218112834

The unit tests do only `import hoomd` and use the fully qualified namespace.

@tpaul-softmat how are you installing HOOMD and azplugins? I think I would need to see some more explicit commands before I can be helpful.

**Reply by tpaul-softmat** [reply]
@mphoward Thanks for your response.

I installed `HOOMD` using `conda`:
`$ conda install -c conda-forge hoomd=4.8.2
`
For the installation of `azplugins`, I followed the instructions given in the `azplugins` [GitHub](https://github.com/mphowardlab/azplugins) page:
```
$ git clone https://github.com/mphowardlab/azplugins
$ cmake -B build/azplugins -S azplugins
$ cmake --build build/azplugins
$ cmake --install build/azplugins
```

**Reply by mphoward** [reply]
Thanks! Are you then also running the pytest command from the same location as the last step?

**Reply by tpaul-softmat** [reply]
Yes, at the same location, with the `--pyargs` flag.
To test HOOMD: `$ python3 -m pytest --pyargs hoomd`
To test azplugins: `$ python3 -m pytest --pyargs hoomd.azplugins`
Both are returning the same error.

**Reply by joaander** [reply, marked as answer]
It is possible that there are ABI incompatibilities if your compilation environment does not match what conda-forge uses to build hoomd-blue.

Try uninstalling hoomd-blue and building it from source as well. You will then need to clear the azplugins build/ directory and rebuild it as well.

**Reply by tpaul-softmat** [reply]
Many thanks for the suggestion.

I tried uninstalling `hoomd-blue` and building it from the source. After the installation, tests for `hoomd` were completed successfully. Next, I tried to reinstall `azplugins`, which gave me the following errors in the `cmake --build build/azplugins` step:
```
Consolidate compiler generated dependencies of target _azplugins
[  4%] Building CXX object src/CMakeFiles/_azplugins.dir/export_AnisoPotentialPairTwoPatchMorse.cc.o
In file included from /home/tanay/build/azplugins/src/export_AnisoPotentialPairTwoPatchMorse.cc:9:
/home/tanay/anaconda3/lib/python3.11/site-packages/hoomd/include/hoomd/md/AnisoPotentialPair.h: In instantiation of ‘void hoomd::md::detail::export_AnisoPotentialPair(pybind11::module&, const string&) [with T = hoomd::azplugins::detail::AnisoPairEvaluatorTwoPatchMorse; pybind11::module = pybind11::module_; std::string = std::__cxx11::basic_string<char>]’:
/home/tanay/build/azplugins/src/export_AnisoPotentialPairTwoPatchMorse.cc:24:66:   required from here
/home/tanay/anaconda3/lib/python3.11/site-packages/hoomd/include/hoomd/md/AnisoPotentialPair.h:708:43: error: ‘getShapeParamName’ is not a member of ‘hoomd::azplugins::detail::AnisoPairEvaluatorTwoPatchMorse’
  708 |         .def(("set" + T::getShapeParamName()).c_str(), &AnisoPotentialPair<T>::setShapePython)
      |                       ~~~~~~~~~~~~~~~~~~~~^~
/home/tanay/anaconda3/lib/python3.11/site-packages/hoomd/include/hoomd/md/AnisoPotentialPair.h:709:43: error: ‘getShapeParamName’ is not a member of ‘hoomd::azplugins::detail::AnisoPairEvaluatorTwoPatchMorse’
  709 |         .def(("get" + T::getShapeParamName()).c_str(), &AnisoPotentialPair<T>::getShapePython)
      |                       ~~~~~~~~~~~~~~~~~~~~^~
gmake[2]: *** [src/CMakeFiles/_azplugins.dir/build.make:216: src/CMakeFiles/_azplugins.dir/export_AnisoPotentialPairTwoPatchMorse.cc.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:117: src/CMakeFiles/_azplugins.dir/all] Error 2
gmake: *** [Makefile:136: all] Error 2

```
I think this happened because of a version mismatch between `hoomd` and `azplugins`. When I built `hoomd` by cloning that from the GitHub repository with the command `$ git clone --recursive https://github.com/glotzerlab/hoomd-blue`, automatically the latest version, which I believe is 4.9.1, is downloaded. However, `azplugins` works with as latest as `hoomd-4.8.2`. If my guess is correct, please let me know how to install `hoomd-4.8.2` or older versions by building it from the source. Also, please let me know how to uninstall the already installed versions of `hoomd` before reinstalling an older version if needed.

Thank you very much for your constant support.

**Reply by mphoward** [reply]
You can checkout the right tag after you clone
```
cd hoomd-blue
git checkout v4.8.2
```
We haven't bumped azplugins to test against 4.9.1, so we will have a look at that.

**Reply by tpaul-softmat** [reply]
Many thanks for the help.

This worked. Tests for both `hoomd` and `azplugins` are completed successfully. So, the solution is to build `hoomd` first [from the source](https://hoomd-blue.readthedocs.io/en/v4.8.2/building.html) by [checking out the right version](https://github.com/glotzerlab/hoomd-blue/discussions/1930#discussioncomment-11218096) compatible with `azplugins` after cloning from GitHub and then to add [`azplugins`](https://github.com/mphowardlab/azplugins) to `hoomd`.

Thank you, @mphoward, @joaander, and @tcmoore3, for all of your valuable suggestions and help.

**Reply by joaander** [reply]
@mphoward have you considered releasing azplugins on conda-forge? That way, users could install binaries directly.

**Reply by mphoward** [reply]
We have discussed doing that for a long time but didn't have time to prepare a recipe. We'd be happy to try if you had an example or template we could follow for a package that follows the component template. Do you have anything you can share?

**Reply by joaander** [reply]
I don't currently maintain any separate components on conda-forge. If I were to create one, I would base it on HOOMD-blue's own recipe: https://github.com/conda-forge/hoomd-feedstock/tree/main/recipe.

**Reply by mphoward** [reply]
OK thanks! I'm going to open a separate discussion thread because I have some other questions about this.
